File: gof-design-patterns/srcs/P01_Singleton/SingletonExample.java
package P01_Singleton;

public class SingletonExample {
    public static void main(String[] args) {
        // 通常のnewはできない（コンパイルエラー）
        // Singleton s = new Singleton(); // エラー！
        
        // getInstance()メソッドでインスタンスを取得
        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();
        
        // 同じインスタンスか確認
        System.out.println("s1 == s2: " + (s1 == s2)); // true
        System.out.println("s1.equals(s2): " + s1.equals(s2)); // true
        
        s1.doSomething();
    }
}

File: gof-design-patterns/typescripts/P01_Singleton/SingletonExample.ts
import Singleton from './Singleton';

// 通常のnewはできない（コンパイルエラー）
// const s = new Singleton(); // エラー！

// getInstance()メソッドでインスタンスを取得
const s1 = Singleton.getInstance();
const s2 = Singleton.getInstance();

// 同じインスタンスか確認
console.log(`s1 === s2: ${s1 === s2}`); // true
console.log(`s1 == s2: ${s1 == s2}`); // true

s1.doSomething();
